### DISCUSSION TITLE
CMake?

### DISCUSSION THREAD

**Discussion by henricj** [adplug/adplug#141] [Ideas]
Any thoughts of regarding having [CMake](https://cmake.org/) build files for adplug?  A project I'm working on builds adplug as a [submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules) for both Linux x64 and Windows x86/x64/ARM64 with [this](https://github.com/henricj/dunelegacy/tree/modernize/external/adplug) setup and I was thinking of adding the tests as well.  It would not be much more work for me to do so inside the adplug directory and providing a PR, instead of in a parent directory as the project is currently doing.  CMake works well on both Windows and Linux, even Visual Studio [natively supports CMake builds](https://learn.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio) these days.  See, for example, my project's [build instructions](https://github.com/henricj/dunelegacy/blob/modernize/README.md#building).

**Comment by Malvineous** [comment]
I'm not against the idea, but it would need someone to continue to maintain it on an ongoing basis.  If you're willing to be involved in the project for at least the next few years, and ensure the CMake files are kept up to date (by fixing issues and giving feedback on any PRs that affect the build system) then it would be fine, but we'd have to remove CMake support again if ever it got too out of date and nobody was willing or able to update it.

Changes that affect the build system are not that frequent so it won't be a lot of work, I only mention it because it probably won't be worth the effort to add it now, if nobody will be around to maintain it and we end up removing it again in a year.

**Comment by henricj** [comment]
I should be able to look at CMake-related build issues for the foreseeable future (at least, within a couple days or so).  That applies to Windows and Linux (or FreeBSD); I don't have much in the way of Mac development experience or a macOS device to test on.

As far as ongoing maintenance is concerned, the only things likely to change are the list of source files to build and, perhaps, the targeted version of the C++ language.

I'll put together a first go at a PR and see what you think (by next week or so, unless something pops up).

**Comment by Malvineous** [comment]
There are automated build scripts in the `.github/workflows/` folder, so I would imagine adding some additional workflows to also test the CMake build process (without interfering with the current build process) would be a good idea, that way it can be tested on Linux and MacOS automatically.  I imagine doing the same thing for `appveyor.yml` would allow automated build testing on Windows too.

There wouldn't be a need to deploy the built image (that's already happening with the existing build process) but you could add some additional workflows/tests that just fail if a CMake build fails.

**Comment by henricj** [comment]
We can [build for Windows on GitHub](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners) also.  Ideally, a single yml with a build [matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) for `macos-latest`, `ubuntu-latest` and `windows-latest` should "just work."  Although, there can be packaging and other issues (e.g., Windows really has three flavors these days: x86, x64, and ARM64) that make it cleaner to have separate yml files for each instead of filling one file with per-platform conditionals.

**Comment by henricj** [comment]
Since [_adplug_](https://github.com/henricj/adplug/) needs [_libbinio_](https://github.com/henricj/libbinio), I started there.  It is not done, but GitHub CI builds for macOS, Windows, and Ubuntu [actually do work](https://github.com/henricj/libbinio/actions/runs/3727074832).  The [workflow](https://github.com/henricj/libbinio/blob/cmake_build/.github/workflows/codeql.yml) scans the code with [CodeQL](https://codeql.github.com/) in addition to building the library.  The [resulting code scan report](https://github.com/henricj/libbinio/security/code-scanning) does nag about `long double` vs `double` (e.g., "[Multiplication result converted to larger type](https://github.com/henricj/libbinio/security/code-scanning/4)").  Right now, _libbinio_ builds stand-alone (on all three platforms) and as a [subdirectory](https://cmake.org/cmake/help/latest/command/add_subdirectory.html) (on Windows, at least).  More _libbinio_ and _adplug_ changes are needed to get _adplug_ to build stand-alone without something else already having installed _libbinio_ where CMake can find it, but it does work as subdirectory.

Changes to _libbinio_ are not needed for the proposed _adplug_ changes to be useful, but it can make development simpler.

**Comment by Malvineous** [comment]
I don't have access to see the code scan report, but I presumed you'd be adding CMake support to libbinio as well to make the whole process CMake-friendly.

I'm happy to merge in the libbinio CMake changes first, so that when you open a PR for Adplug the automatic CI run will pass, but it would be good if you could wait until the whole thing is complete on your end across both libraries before opening the first PR.

**Comment by henricj** [comment]
Come to think of it, making the security scan private makes sense.  Providing a searchable public database of possible security vulnerabilities that have not yet been fixed has some obvious drawbacks.  I sent an invite to that fork, mostly out of curiosity to see if that would let someone view the security scan (alternately, pushing the `cmake_build` branch to a repo you own should work).

As for PRs:  First I want to get both `libbinio` and `adplug` to the point where they work stand-alone and as [subdirectories](https://cmake.org/cmake/help/latest/command/add_subdirectory.html) (tested with my [`dunelegacy`](https://github.com/henricj/dunelegacy) project).  Also, `adplug` should work both with a system-installed `libbinio` (say, from an `apt-get` on Ubuntu) and with something the user built themselves.  Most of that should hopefully happen this week sometime.

There are a couple other changes that aren't directly related to the CMake stuff.  A couple are trivial additions: adding [`.editorconfig`](https://github.com/henricj/adplug/commit/880efc104c47cee07053670efcd11d7f3edee728) files and listing some more Visual Studio gunk in [`.gitignore`](https://github.com/henricj/libbinio/commit/5cbc1c49a434609ff4cf685af8e270c6df81ccb4).  If they are of interest, either snipe them or let me know and I can create separate PRs.  There is also anther more [significant change](https://github.com/henricj/libbinio/commit/b11cb495eb2c0512dda640f07386255624f221f3) where git is used to set the project version and generate a private header file like so (from Ubuntu):

```c
#define GITVERSION_LONG "69f8dc54294349c60852f06972efbbd3cf02aaeb"
#define GITVERSION_TIME "Sun Dec 18 16:25:04 2022 -0700"
#define GITVERSION_REPO_URL "https://github.com/henricj/libbinio.git"
#define GITVERSION_REPO_BRANCH "cmake_build"
#define GITVERSION_DESCRIBE "libbinio-1.5-17-g69f8dc5"
#define GITVERSION_MASTER_MERGE_BASE "e2f8d50c53102c618d675c3310e09a0e0bdf49cd"
#define GITVERSION_DESCRIBE_MERGE_BASE "libbinio-1.5-9-ge2f8d50"
```
If git can't be found or if the build is not done from a git repo, then it should be a no-op (apart from a warning in the build log).  The "*_MERGE_BASE" values might not be relevant here.

On the `dunelegacy` project, I've found both [clang-tidy](https://clang.llvm.org/extra/clang-tidy/) and [clang-format](https://clang.llvm.org/docs/ClangFormat.html) to be useful.  The former's fixes need to be run manually since it sometimes garbles the source and the latter often changes even on minor LLVM version changes, but even so, the diagnostics are useful and it is nice to get the code into a uniform style (`libbinio` and `adplug` look to be in pretty good shape already, but `dunelegacy` was a mess).  The configuration files used for `dunelegacy` look like this: [`.clang-tidy`](https://github.com/henricj/dunelegacy/blob/modernize/.clang-tidy) and [`.clang-format`](https://github.com/henricj/dunelegacy/blob/modernize/.clang-format).
